### DISCUSSION TITLE
Floorplan / datacenter grid plugin

### DISCUSSION THREAD

**Discussion by xkilian** [netbox-community/netbox#5844] [Ideas]
@bellwood , @ryanmerolle, @ndom91, @tb-killa,  @candlerb 
Anyone interested in rebooting this question as a plugin, put a first version in github so that everyone interested can contribute in a community driven fashion. This is one of the basic helpful features that should be implemented in netbox. A floorplan with racks and non racked devices. 

Issues that need to be addressed:

- Floorplan/ datacenter grid position needs to be in the data model (x,y) for racked devices, (x,y,z) for unracked devices, (x,y,z position for racks themselves.Might be data that is included in #4971 
- ability to add devices to rack-groups so they can be positioned in the room (ex. Inline Row Cooling unit, Batteries, Ventilation modules, etc.) #4971 

Features: 
- calculate and display rack occupancy, claulcate and display max free contiguous number of Us in racksç
- calculate and display btu per rack, #5207 , per room
- calculate and display PDU occupancy and % free amps?
- support layers (eletrical, cooling, occupancy)
- designation on floorplan cool/hot aisles
- Move around or reposition unracked devices on the floorplan
- Move around or reposition racks on the floorplan

Exclusions:

- I do not see this as a method for displaying 2d front-back vews of racks (Netbox already handles this)
- I do not see this as a help in physical plant cabling layout help
- I do not see this as a method to move racked equipment around

This is simply a call to see if a first draft can be done and get things rolling, as it will not be done within Netbox directly and can be handled well as a plugin I would think. Obviously #4971 is an important piece that is missing in how the implementation can move along.

@bellwood You had some time of prototype underway, do you want to share this on github as a starting point or simply discuss design before committing something to github?

**Comment by ndom91** [comment]
Yeah I think keeping it as simple as possible for an MVP is a good goal though.

How about, simply adding an (x,y) position field to the Rack model, and then adding the room / display logic.

We can add various layers (cooling, electrical) as well as additional devices like cooling units in rack groups, etc. later on imo.

Other notes that go through my head initially:
- rack devices already have an z value, i.e. height in the rack. If we give the rack an x,y - the devices will obviously be in that rack and can adopt the rack's x and y coordinate and use the unit # aka the rack height as their z value. 
- rack occupancy % is already calculated and displayed in the rack view table, we can just grab that value
- pdu occupancy % / free amps can already be calculated and displayed as well with existing power panel features

So to summarize, I think a lot of this is already in Netbox. We'd just need to add a new view where one can create a grid to act as the floorplan, and then place racks on it, i.e. give them x,y coordinates on that grid. 

Theres also no hard and fast rule about how exactly to organise your netbox sites, racks, rack groups, etc. but I think most people use rack groups to approximate rooms, is that what you guys do? Currently, in terms of the code, I think rack groups in netbox don't have a model themselves, but are simply an key under which other models (Racks) can be grouped, so to speak. So we'd need to convert it into an actual "thing" / model so that it can have its own detail page so that we can add fields for the grid size, shape, maybe also comments, etc.

What do you guys think?

**Reply by xkilian** [reply]
A canvas view
A management view for each equipement to manage its size, location and maybe a way to do quick edits for its electrical consumption.
Add a x,y,z to Locations model. (Which replaces the Rack Groups.)
z is required as devices may be wall mounted. z would represent the height of the bottom of the equipement. Maybe this is too advanced for a first version.
An equipement that is displayed in the grid should be sized to scale. Which means that an equipment should have height, width, length information associated.
This also begs the question for vertically super-imposed equipement. Should be allowed and a method should permit to understand that more than one équipement occupies the space.
#4971 is now closed and no longer blocking. Locations now have data models.

The main problem is determining a programmable canvas for displaying the grid.

**Reply by dBitech** [reply]
a good candidate for the programable canvas would be in my opinion https://github.com/jsGraph/mxgraph

**Comment by xkilian** [comment]
Our use and the most common use that I have read is that rack groups are containers and the first level of containers is usually a room. So it fits as you suggest. #4971 should be done in Netbox itself and that should be leveraged to add x,y,z for those devices.

I agree to keeping it simple, but the library that implements de canvas for the floorplan/grid should be flexible enough for better usability.  I expect @bellwood could share his experience with what he used and if it would still be a good fit or suggest something else.
Good idea to associate the model of the floor plan to a rackgroup. Though I would expect it be limited to the first container (room), I do not think rackgroups are commonly used for rows or other intermediate use-cases. Especially with a floorplan that type of intermediate rackgroup does not make much sense.

**Reply by ndom91** [reply]
Yes good call on #4971 - that would fit best in Netbox itself and is sort of a "blocker" for a plugin like this.

I agree the floorplan should be flexible enough to accomodate many different organising principles as rack groups aren't specifically defined as beign for "rooms". I forgot rack gruops can recursively contain themselves, that may be a problem haha. We may have to define a limitation for use - like rack groups can only be 1 or 2 levels deep, i.e. they have to represent rooms in order to use this plugin.

Lets try to brainstorm what blockers there are in Netbox itself to completing something like this, and then we can start coming up with specific units of work, etc, etc.

Like I said, I'm not super familiar with the Netbox codebase but I've dug through some areas before, and im 99% sure that Rack Groups would need to be converted to having a model itself. Currently its just like 'Tenant Groups', for example, an organising option and not a Netbox model itself with its own page with its own fields, etc. 

Otherwise I can't think of anything else off the top of my head that would be a blocker in Netbox. The rest would just be work in the plugin itself - creating a new view with the grid, etc. We'd have to add a field to the rack model to save its location on its rack group's grid. Or maybe duplicate all racks into our own table in the db to save all our fields (x,y,z location, other notes, etc.) so we don't pollute the rack model in netbox itself.

**Reply by rasanentimo** [reply]
Would it be possible to identify the rack group which represents the room/floor with tags for example? This way it would be possible to use nested rack groups without limiting the depth.

**Reply by ndom91** [reply]
Unfortunately rack groups in their current state have no additional fields (tags, comments, etc.). But if we need to change rack groups to be a model itself, I dont see why we couldn't add tags and, for example, only use rack groups tags as `#room` (for example).

**Reply by rasanentimo** [reply]
As #4971 was part of the 2.11 release, how should we proceed?

**Reply by xkilian** [reply]
Data model requirements:

- locations can now contain equipement
- locations can have avec custom fields
- devices can have custom fields
- "racks/cabinets" can have custom fields (i think)
- for a given location, the plugin needs to know if it is a room
- an unracked device can have a custom field with its PX, PY, PZ position
- an unracked device can have a custom field for its volume VX, VY, VZ
- a cabinet can have a custom field with its PX, PY, PZ position
- a cabinet can have a custom field for its volume VX, VY, VZ

The plugin should probably manage in a standardized way rack volumes so that are pre-templated and simply applied to racks in a room. Same for un-racked devices.
The plugin needs to manage the layout/editing of the physical layer. Other layers could be handled in subsequent layers (cooling hot/cold aisles, sensors, AC Units, electrical), (PDU/Transfer switches/Panel/etc.))
The plugin needs to manage displaying meta-data about non-racked devices and racks.
The plugin needs to calculate any missing data that needs to be displayed about racks. (Power efficiency, etc.)
The plugin needs to integrate with the new netbox logging functions
The new custom fields should be displayed in Netbox for devices and racks.

I am not a developper, so I am only organiseing the data for others to be able to run ith it.
Cheers.

**Reply by ndom91** [reply]
Sorry guys I started a new job at a startup a few months ago and have no bandwidth to work on something bigger like this anymore :(

**Reply by ndom91** [reply]
Afaik all assets were turned into models with metadata a while back. So for example, instead of "rack groups" just being a "folder" for racks, it's been renamed "Site groups" (afaict) and has a model itself with its own metadata fields. So we could add a `px`, `py`, `pz`, etc. to that entity now.

Like I said above though, I no longer work at the employer using Netbox and unfortunately don't have time to work on this in my free time :/

**Comment by maxzapp** [comment]
definitely a needed feature

**Reply by deku-m** [reply]
Did anyone ever created a plugin for this?

**Reply by ndom91** [reply]
I don't think theres been any progress here as far as I know

**Comment by dreng** [comment]
The list of wanted features is long. Maybe it helps to simplify the requirements in order to let the project fly.

I created a simple, zoomable and movable Javascript raster view just as a proof of concept. It can take objects and place them in a defined raster. We only need the following information from NetBox: Raster size (corresponds to room dimensions) and position of each object. I didn't care about real dimensions for the sake of simplicity. I have never worked in a datacenter, therefore I couldn't tell what is really important about such a plugin and what is nice to have.

You can try the demo here: https://jsfiddle.net/tmso6y8u/

Anyway the project needs some work to be done. As this should not be a fire and forget project, the question is, how many developers / contributers are willing to maintain this project long term. I could certainly help but I've got enough projects of my own and would only join a team which has enough members. Any Volunteers?

**Reply by ndom91** [reply]
Oh wow this is awesome! Yeah that'd be a great first release and somethign solid to build upon for some of the additional requests.

**Comment by bellwood** [comment]
For our floor plans, I simply used the rack names as DOM IDs and used Flexbox and other CSS to position things as needed.

@dreng's concept is pretty cool too, I may look at that further.

**Comment by dreng** [comment]
@bellwood That sounds interesting aswell. In general, I'd prefer solutions without Javascript. When I thought about my goal and experimented with flexboxes a bit, I found it hard to get what I wanted and the code felt quite bloated. So I gave up very, very quick. That's why I used Javascript. But to be honest, I'm completely unexperienced with flexboxes. May be you'd like to upload a demo of what you created?

Although I wouldn't like to through my development in the bin, this thread shouldn't be a competition. It's better to throw something away which is in an early state than to insist on keeping a worse solution. So let's show and discuss and find out what technique fits best in the long run.

I hope that some experienced networkers and developers will join the discussion.

**Reply by ndom91** [reply]
Without JavaScript your not going to have any sort of click n drag type UI. 

Using flex box you could give the user some basic options, like position all these racks in a row or a column, but that's about it. This feature would be extremely limited imo with only css support. 

Imo giving the user a way to design a raster of their space makes the most sense imo. Then they can click n drag individual racks onto the spots to fill them in so to speak

**Reply by bellwood** [reply]
Unless we add rack groups, my infrastructure assets don't really change much and are cookie-cutter,  so I went with a Flexbox CSS layout that is responsive. 

That said, my floor plan is part of a separate in-house app our team uses, and is generated using the Netbox API.

**Comment by PVM-NL** [comment]
I think the most important thing is to have some kind of raster/image where we place the rack on, could be in first version pretty straight forward. Like have a big raster place the rack reserved on it with name or short name |R1|R2|R3| .
![Naamloos](https://user-images.githubusercontent.com/120473446/211318976-2d5e90e5-5c68-4db2-ab40-714ac600ff91.png)
Just like racks are draw. But we then need to connect the raster to the site or location. On there they need to link to the specific RACK.

The features mention above some are already implemented on rack itself. But can be added through further development.
Im not a programmer so i can only view from the side line. Cant say what technique is the best to provide in this.

**Reply by dreng** [reply]
Have you had a look at the demo I provided a few posts before? It looks quite similar to your mockup.
Regarding the link to rack: Technically speaking, the other way round would be a better solution. Every rack is bound to a site or location. We just need to store, which position in the raster a should have. The dataset for the raster view will then be generated by querying and filtering the database.

**Reply by PVM-NL** [reply]
@dreng Yes, i had. Thats why it was just my suggestion to start with something like your example. Forgot to reply to your comment.
About how to link it, i would agree on what you say.

**Comment by bertdwe** [comment]
I posted below also in [netbox_floorplan](https://github.com/tbotnz/netbox_floorplan) but I believe below is site core functionality.
We use currently OpenDCIM , but looking into migration scenario's.

A typical site, with a floorplan. The background image is an imported plan/drawing. The racks have an overlay, created within the application, clickable and link to the rack or row views, which is a really convenient way of navigating.

![image](https://github.com/netbox-community/netbox/assets/78160087/7558b78d-5b5a-48c0-9bd9-766fea5c195d)


Rack colors change dynamically. Hovering over a rack shows tooltips with rack details.
How this works is explained in more detail in https://www.youtube.com/watch?app=desktop&v=8p-tFL9tJxw
